### PR TITLE
gui: When BIP70 is disabled, get PaymentRequest merchant using string search

### PR DESCRIPTION
**Pull request by achow101**
The merchant name is stored in the X.509 certificate embedded in a PaymentRequest. Use some string searching to locate it so that it can be shown to the user in the transaction details when BIP70 support was not configured.

An additional notice is added to the merchant string that indicates the certificate was not verified. When BIP70 is enabled, the certificate would be verified and the merchant name not shown if the certificate was invalid.

### CODE REVIEW

File: src/qt/transactiondesc.cpp [hunk 1]
```diff
@@ -259,22 +259,45 @@ QString TransactionDesc::toHTML(interfaces::Node& node, interfaces::Wallet& wall
...
         {
+#ifdef ENABLE_BIP70
             PaymentRequestPlus req;
             req.parse(QByteArray::fromRawData(r.second.data(), r.second.size()));
             QString merchant;
             if (req.getMerchant(PaymentServer::getCertStore(), merchant))
                 strHTML += "<b>" + tr("Merchant") + ":</b> " + GUIUtil::HtmlEscape(merchant) + "<br>";
+#else
+            // Search for the supported pki type strings
+            std::string pr = r.second;
```
Comment (laanwj) [inline src/qt/transactiondesc.cpp]: Please, factor this out to a function at least :smile:

**Comment by laanwj** [comment]
Concept ACK

Are there other places where the vendor name is used? For ex. the transaction list?

File: src/qt/transactiondesc.cpp (hunk 1 above)
Comment (achow101) [reply src/qt/transactiondesc.cpp]: Put it into a function and cleaned up this for loop and if block a bit.

**Comment by achow101** [comment]
> Are there other places where the vendor name is used? For ex. the transaction list?

AFAIK, no. @luke-jr might know as he was the one that pointed this out to me.

File: src/qt/transactiondesc.cpp (hunk 1 above)
Comment (laanwj) [reply src/qt/transactiondesc.cpp]: thanks !

**Comment by laanwj** [comment]
> AFAIK, no. luke-jr might know as he was the one that pointed this out to me.

I checked: you're right, the only place outside of paymentrequestplus / paymentserver and the tests where the `getMerchant` method is used is here.

**Comment by luke-jr** [comment]
Concept ACK

**Comment by laanwj** [comment]
Now, we need to find someone that has actual paid payment requests in their wallet to test this :smile:

**Comment by harding** [comment]
Tested fc295e42071a82955c35d90792a9cbd127ca2859 works for me on transactions I made via BIP70 back in 2015.

**Comment by fanquake** [comment]
Thanks for doing some "real world" testing @harding.

**Comment by Sjors** [comment]
Concept ACK.

Tested fc295e4 on macOS 10.14 without BIP70 support by looking at some BitPay transactions from late 2018. It shows the merchant with "(Certificate was not verified)"

However when I test with BIP70 support it doesn't show the merchant. That doesn't work on master either.

A test case for `GetPaymentRequestMerchant` would be nice.

**Comment by achow101** [comment]
> However when I test with BIP70 support it doesn't show the merchant. That doesn't work on master either.

Presumably the certificate is no longer valid. I believe they have a new certificate since earlier this year, so the old one expired and thus the merchant name isn't being shown now.

File: src/qt/transactiondesc.cpp [hunk 2]
```diff
@@ -255,26 +285,34 @@ QString TransactionDesc::toHTML(interfaces::Node& node, interfaces::Wallet& wall
     strHTML += "<b>" + tr("Output index") + ":</b> " + QString::number(rec->getOutputIndex()) + "<br>";
 
     // Message from normal bitcoin:URI (bitcoin:123...?message=example)
-    for (const std::pair<std::string, std::string>& r : orderForm)
+    for (const std::pair<std::string, std::string>& r : orderForm) {
```
Comment (luke-jr) [inline src/qt/transactiondesc.cpp]: This will change Message,Message,Merchant,Merchant into Message,Merchant,Message,Merchant.

Is that intentional? Probably should be in a separate commit...

Comment (achow101) [reply src/qt/transactiondesc.cpp]: Does it really matter that much that the order is changed?

The only reason I changed it is because it looked funny that there were two loops in a row over the same vector.

**Comment by wiranphowan** [comment]
wiranthphothiwan@gmail.com

ในวันที่ ส. 21 ก.ย. 2019 18:03 Andrew Chow <notifications@github.com>
เขียนว่า:

File: src/qt/transactiondesc.cpp [hunk 3]
```diff
@@ -49,6 +49,36 @@ QString TransactionDesc::FormatTxStatus(const interfaces::WalletTx& wtx, const i
     }
 }
 
+#ifndef ENABLE_BIP70
+// Takes an encoded PaymentRequst as a string and tries to find the Common Name of the X.509 certificate
```
Comment (practicalswift) [inline src/qt/transactiondesc.cpp]: Should be `PaymentRequest` :)

**Comment by laanwj** [comment]
@achow101 Can you do the small comment correction please ^^

If not, this will be merged nevertheless in the course of today or tomorrow before the 0.19 split-off. It's more desirable to have this with a comment typo than not at all.

File: src/qt/transactiondesc.cpp (hunk 3 above)
Comment (achow101) [reply src/qt/transactiondesc.cpp]: Done

**Comment by laanwj** [comment]
ACK 85973bcc44f60fe3bbc952557ebf578dd4c475d2

**Comment by dooglus** [comment]
I have a couple of payments to bitpay in my wallet. In Bitcoin Core 0.18.1 the merchant information wasn't displayed at all for either of them. When I viewed the transaction details I would see this in the debug.log:

    PaymentRequestPlus::getMerchant: Payment request: certificate expired or not yet active

The certificate was valid from 2017-03-23 18:22:00.000 UTC to 2019-04-25 19:11:00.000 UTC, and so could no longer be validated, and as a result Core wasn't showing any "Merchant:" information at all.

In Bitcoin Core 0.19 I see:

    Merchant: bitpay.com (Certificate was not verified)

which is infinitely better.

**Comment by laanwj** [comment]
@dooglus Thanks for checking. Yes it was a bug that the certificate-checking for transactions used the current date, instead of the date of the transaction (also, verifying it once at payment would have been good enough, no need to do so every time details are viewed).
